#4547: Refactor vmm builder code to simplify logic that creates the microVM to boot

tommady opened the pull request:
fix(4547): Refactor vmm builder code to simplify logic

## Changes
Refactoring Vmm builder code

## Reason
close issue #4547 

...

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [x] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [x] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [ ] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [x] If a specific issue led to this PR, this PR closes the issue.
- [ ] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [x] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [ ] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -730,13 +707,39 @@ fn attach_legacy_devices_aarch64(
+        vcpus.push(vcpu);
+    }
+    // Make stdout non blocking.
+    set_stdout_nonblocking();
```
This seems to belong to the serial device init, so can be moved back to pio device manager init code.

tommady replied:
The aarch64::attach_legacy_devices function also relies on this functionality. Moving it into the PIO device manager would make implementation significantly more challenging.

ShadowCurse replied:
How about move this into the `setup_serial_device`? I see that we always call `set_stdout_nonblocking` before attaching the serial.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -148,19 +148,342 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+pub mod aarch64 {
+    use super::*;
+
+    fn attach_legacy_devices_aarch64(
```
The `aarch64` suffix can be dropped.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -308,11 +644,23 @@ pub fn build_microvm_for_boot(
 
     #[cfg(target_arch = "aarch64")]
-    attach_legacy_devices_aarch64(event_manager, &mut vmm, &mut boot_cmdline).map_err(Internal)?;
+    aarch::attach_legacy_devices_aarch64(event_manager, &mut vmm, &mut boot_cmdline)
```
`s/aarch/aarch64` in the module name

ShadowCurse replied:
I meant `aarch::attach_legacy_devices` -> `aarch64::attach_legacy_devices` since `aarch64` is the module name.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -148,19 +148,342 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+
+        let vcpu_config = VcpuConfig {
+            vcpu_count: vm_config.vcpu_count,
+            smt: vm_config.smt,
```
hmm.. considering smt does not exist on aarch64 anyway, maybe we can skip it here or always set it to `false`?

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -168,7 +168,9 @@ pub mod aarch64 {
         if cmdline_contains_console {
             // Make stdout non-blocking.
             set_stdout_nonblocking();
-            let serial = setup_serial_device(event_manager, std::io::stdin(), std::io::stdout())?;
```
The last "fix: Refactor vmm builder code to simplify logic" should be squashed with previous one.

ShadowCurse reviewed (commented):
Hi @tommady, thx for doing the PR. Overall changes seems solid, but I left some small notes here and there. Also can you change names of the commits to `refactor(builder): ...`. This is usually how we treat refactor commits.

tommady commented:
Hi @ShadowCurse 

I’ve addressed your comments, with one note regarding the set_stdout_nonblocking function. Please review and provide your guidance when you have a moment.

Thank you!

ShadowCurse reviewed (commented):
Now you need to make sure the changes you made pass our CI. You can look at the `buildkite/firecracker-pr` line at the bottom of the PR and click `details` to look at the CI status. Start with making sure changes do compile. I can see that on x86_86 only errors are absence of doc comments for public functions. For aarch64, well, it does not even compile. If you don't have access to arm system, you can try to use [cross](https://github.com/cross-rs/cross) to cross compile. 
After everything compiles, please move all changes to corresponding commits.
Also please rebase the PR so it only includes your commits.

tommady commented:
Hi @ShadowCurse 

Thank you for your patience and thorough review—it has been an incredible learning experience for me!
I’ve revisited your feedback and ensured that the entire architecture compiles cleanly, with no documentation warnings.
Please take a look whenever you have the time. 😊

ShadowCurse reviewed (commented):
Now that it at least compiles, you need to make sure it also does function as expected. If you look at the Buildkite pipelines output: https://buildkite.com/firecracker/firecracker-pr/builds/11856#0193640b-6e7c-4f49-bbcf-cb60c1c946ef/62-392 you will see that there are some unit tests that fail in the same place `src/vmm/src/builder.rs:519:51`. Quick look around and it seems the issue is that when `pio_device_manager` registers itself, it registers interrupts for a VM, but the interrupt controller is not yet created at that point. So you need to either move creation of interrupt_controller before `pio_device_manager` creation, or move this `.register_device(..)` call. 

You can reproduce all of this errors if you run CI locally. You can find exact commands CI is running looking at the very beginning of the logs like this: https://buildkite.com/firecracker/firecracker-pr/builds/11856#0193640b-6e7c-4f49-bbcf-cb60c1c946ef/62-63

 Start by running unit tests. There are a couple of ways to do this:
- `./tools/devtool test -- integration_tests/build/test_unittests.py` - this will use the docker container to compile and run tests
- Compile test with `cargo test -p vmm --no-run` and then run with `sudo ./build/cargo_target/debug/deps/vmm-....` (there will be a full path at the end of `cargo test`) With this method some network related unit tests will fail most likely, but you can skip those.

Also you can make sure your changes are sound by to simply trying to launch a VM with them. We have a doc for this: https://github.com/firecracker-microvm/firecracker/blob/main/docs/getting-started.md (in the section about getting a rootfs you can skip the ssh key creation if you don't want to use/test network)

tommady commented:
Hi @ShadowCurse,
Thank you once again for your patience and for crafting such a detailed guide! 🙏
I’ve fixed the unit test as per your suggestions and ran the command:
```bash
./tools/devtool test -- integration_tests/build/test_unittests.py
...
integration_tests/build/test_unittests.py::test_unittests PASSED                                                                                         [ 50%]
integration_tests/build/test_unittests.py::test_benchmarks_compile PASSED                                                                                [100%]
```
All tests passed successfully! 🎉

When you have a moment, please review it again.

ShadowCurse commented:
With x86_64 version done, now you need to make sure aarhc64 also compiles/works. If you look at the BK run again: https://buildkite.com/firecracker/firecracker-pr/builds/11886#01936ce9-9b52-4459-ac0a-f8cfbe7a3690/62-624, https://buildkite.com/firecracker/firecracker-pr/builds/11886#01936ce9-9b62-4713-b1eb-33926eacc8b0/62-882 you will see some compilation errors in unit tests and CI test for serial port in a restored VM failing. As I mentioned before, you can use [cross](https://github.com/cross-rs/cross) to cross compile FC to aarch64. It might even allow some subset of unit tests to be run locally. Or you can use qemu to create an emulated aarch64 VM and test your code there.

tommady commented:
Oh, I see… I assumed the devtool test would handle all platform testing in one sweep. Let me double-check again.

ShadowCurse commented on src/vmm/src/device_manager/persist.rs:
```diff
@@ -418,13 +418,11 @@ impl<'a> Persist<'a> for MMIODeviceManager {
-                        std::io::stdin(),
-                        std::io::stdout(),
-                    )?;
+                    let serial = setup_serial_device(std::io::stdin(), std::io::stdout())?;
```
The `event_manager.add_subscriber` call is missing here and causing CI test to fail.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -148,26 +148,344 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+            vcpus.push(vcpu);
+        }
+
+        setup_interrupt_controller(vm, vcpu_count)?;
```
Let's move creation of interrupt controller out of `create_vcpus` and into the `create_vmm_and_vcpus`.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -184,23 +502,19 @@ fn create_vmm_and_vcpus(
             .map_err(VmmError::EventFd)
             .map_err(Internal)?;
 
+        x86_64::setup_interrupt_controller(&mut vm).map_err(Internal)?;
```
Here we can also move creation of interrupt controller into the `create_vmm_and_vcpus`. But with it we need to move the `pio_dev_mgr.register_devices(vm.fd()).unwrap();` line there as well.

ShadowCurse reviewed (commented):
Left couple of comments and as I think, the reason for CI failures. After addressing everything please rebase PR on top of main so only your commits are in here.

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -148,26 +148,372 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+            kvm_capabilities,
+        )?;
+
+        let pio_device_manager = {
+            // Serial device setup.
+            let serial_device = setup_serial_device(std::io::stdin(), io::stdout())
+                .map_err(StartMicrovmError::Internal)?;
+
+            // x86_64 uses the i8042 reset event as the Vmm exit event.
+            let reset_evt = vmm
+                .vcpus_exit_evt
+                .try_clone()
+                .map_err(VmmError::EventFd)
+                .map_err(StartMicrovmError::Internal)?;
+
+            setup_interrupt_controller(&mut vmm.vm).map_err(StartMicrovmError::Internal)?;
+
+            // create pio dev manager with legacy devices
+            let pio_device_manager = {
+                // TODO Remove these unwraps.
+                let mut pio_dev_mgr = PortIODeviceManager::new(serial_device, reset_evt).unwrap();
+                pio_dev_mgr.register_devices(vmm.vm.fd()).unwrap();
+                pio_dev_mgr
+            };
+
+            pio_device_manager
+        };
```
The `pio_device_manager` itself should be constructed where it was previously in the `build_vmm` and should not be `Optional` as there is no situation where x86 vmm does not have one. Only things you need to keep here are 2 lines:
```
setup_interrupt_controller(&mut vmm.vm).map_err(StartMicrovmError::Internal)?;
and 
pio_dev_mgr.register_devices(vmm.vm.fd()).unwrap();
```

tommady replied:
thanks for teaching me this 🙇🏻

ShadowCurse commented on src/vmm/src/builder.rs:
```diff
@@ -148,26 +148,372 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
     // Build custom CPU config if a custom template is provided.
+    //
+    // allow unused_mut for the aarch64 platform.
+    #[allow(unused_mut)]
```
Since this `mut` was only needed for `setup_interrupt_controller` call, it can be removed.

ShadowCurse commented:
Alright, CI functional and performance tests are green. The last thing is to make commits look better. As I mentioned before you need to rebase PR on top of main branch so only your commits are visible. This should be as simple as updating you fork of Firecracker and running `git rebase origin/main`. After that update commits with meaningful titles and messages and squash some of them if necessary.

ShadowCurse commented:
Hi @tommady, are you still interested in finishing this PR? Otherwise I can take over and bring it to completion.

tommady commented:
Apologies for missing this PR, and thanks for taking care of it.
